diff --git a/CMakeLists.txt b/CMakeLists.txt
index c47bb0126..bfeb0b7cc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -321,7 +321,9 @@ if(TRITON_BUILD_PYTHON_MODULE)
     set_target_properties(triton PROPERTIES SUFFIX ".pyd")
     set_target_properties(triton PROPERTIES PREFIX "lib")
   else()
-    target_link_libraries(triton PRIVATE z)
+    # target_link_libraries(triton PRIVATE z)
+    find_package(ZLIB)
+    target_link_libraries(triton PRIVATE ZLIB::ZLIB)
   endif()
   target_link_options(triton PRIVATE ${LLVM_LDFLAGS})

diff --git a/setup.py b/setup.py
index d8e086029..2597c859d 100644
--- a/setup.py
+++ b/setup.py
@@ -405,7 +405,7 @@ class CMakeBuild(build_ext):
         build_ext.finalize_options(self)

     def run(self):
-        download_and_copy_dependencies()
+        # download_and_copy_dependencies()

         try:
             out = subprocess.check_output(["cmake", "--version"])
@@ -845,7 +845,7 @@ setup(
     package_dir=dict(get_package_dirs()),
     entry_points=get_entry_points(),
     include_package_data=True,
-    ext_modules=[CMakeExtension("triton", "triton/_C/")],
+    # ext_modules=[CMakeExtension("triton", "triton/_C/")],
     cmdclass={
         "bdist_wheel": plugin_bdist_wheel,
         "build_ext": CMakeBuild,
